gateway/src/api/cache_layer.rs

```diff
@@ -0,0 +1,275 @@
+
+                let public_key_fn = self.public_key_fn.clone();
+                let cache_manager = self.cache_manager.clone();
+
+                return Box::pin(async move {
+                    let response: Response = future.await?;
+
+                    // Return response directly if it failed.
+                    if response.status() != StatusCode::OK {
+                        return Ok(response);
+                    }
+
+                    // We'll re-use the parts in the response if all goes well.
+                    let (parts, body) = response.into_parts();
+
+                    let body = match hyper::body::to_bytes(body).await {
+                        Ok(body) => body,
+                        Err(error) => {
+                            error!(
+                                error = &error as &dyn std::error::Error,
+                                "failed to get response body"
+                            );
+
+                            return Ok(Response::builder()
+                                .status(StatusCode::INTERNAL_SERVER_ERROR)
+                                .body(Default::default())
+                                .unwrap());
+                        }
+                    };
+
+                    let value: Value = match serde_json::from_slice(&body) {
+                        Ok(value) => value,
+                        Err(error) => {
+                            error!(
+                                error = &error as &dyn std::error::Error,
+                                "response body is malformed"
+                            );
+
+                            return Ok(Response::builder()
+                                .status(StatusCode::BAD_REQUEST)
+                                .body(Default::default())
+                                .unwrap());
+                        }
+                    };
+
+                    let Some(jwt) = value["token"].as_str() else {
+                        error!("response json is missing 'token' key");
+
+                        return Ok(Response::builder()
+                                .status(StatusCode::UNAUTHORIZED)
+                                .body(Default::default())
+                                .unwrap());
+                    };
+
+                    let public_key = match public_key_fn.public_key().await {
+                        Ok(public_key) => public_key,
+                        Err(error) => {
+                            error!(
+                                error = &error as &dyn std::error::Error,
+                                "failed to get public key"
+                            );
+
+                            return Ok(Response::builder()
+                                .status(StatusCode::SERVICE_UNAVAILABLE)
+                                .body(Default::default())
+                                .unwrap());
+                        }
+                    };
```

Comment:
I wonder if we can get the middle part of the jwt here and then base64 decode it. Then finally convert it to a `serde_json::Value` to read the `exp`. This way we won't have to convert it to a claim and won't need the public key function.

An example of how we could get the `exp`
https://github.com/shuttle-hq/shuttle/blob/e8536e881e33a93491750e085615b71665da988d/common/src/backends/auth.rs#L779-L785

Reply:
Because of all the matches this will produce, you might want to make a function that takes the body (line 156) as an argument and will calculate the duration (line 237). Then have this function's return type as `Result<Duration, StatusCode>` and you'll be able to replace the `match`es with `?`

Reply:
That's a good idea, I'll do that. Thanks!